server/src/main/java/org/elasticsearch/search/aggregations/InternalAggregations.java

```diff
@@ -121,11 +148,27 @@ public void readFrom(StreamInput in) throws IOException {
         if (aggregations.isEmpty()) {
             aggregationsAsMap = emptyMap();
         }
+        //TODO update version after backport
+        if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
+            if (in.readBoolean()) {
+                this.topLevelPipelineAggregators = in.readList(
+                    stream -> (SiblingPipelineAggregator)in.readNamedWriteable(PipelineAggregator.class));
+            }
+        }
     }
 
     @Override
     @SuppressWarnings("unchecked")
     public void writeTo(StreamOutput out) throws IOException {
         out.writeNamedWriteableList((List<InternalAggregation>)aggregations);
+        //TODO update version after backport
+        if (out.getVersion().onOrAfter(Version.V_8_0_0)) {
+            if (topLevelPipelineAggregators == null) {
```

Comment:
can we rely on an empty list if the aggregations are completely reduced ? This way we don't need the boolean and can call the write list directly.

Reply:
yes, I initially did not do it but I just realized that I can. I was worried about cases with CCS where we receive from e.g. 6.6 and then write the same object to e.g. 7.x. I just need to set the list to an empty one in that case too which removes the need for the list to be nullable 100%.